optimum/habana/transformers/models/qwen2_vl/modeling_qwen2_vl.py

```diff
@@ -0,0 +1,763 @@
+    Qwen2VLFlashAttention2,
+    Qwen2VLForConditionalGeneration,
+    Qwen2VLModel,
+    Qwen2VLSdpaAttention,
+    Qwen2VLVisionBlock,
+    VisionAttention,
+    VisionFlashAttention2,
+    VisionSdpaAttention,
+    _prepare_4d_causal_attention_mask_with_cache_position,
+    apply_multimodal_rotary_pos_emb,
+    apply_rotary_pos_emb_vision,
+    repeat_kv,
+)
+from transformers.utils import logging
+
+
+try:
+    from habana_frameworks.torch.hpex.kernels import FusedSDPA
+except ImportError:
+    print("Not using HPU fused scaled dot-product attention kernel.")
+    FusedSDPA = None
+
+logger = logging.get_logger(__name__)
+
+
+class ModuleFusedSDPA(torch.nn.Module):
+    def __init__(self, fusedSDPA):
+        super().__init__()
+        self._hpu_kernel_fsdpa = fusedSDPA
+
+    def forward(self, query, key, value, attn_mask, dropout_p, is_casual, scale, softmax_mode):
+        return self._hpu_kernel_fsdpa.apply(query, key, value, attn_mask, dropout_p, is_casual, scale, softmax_mode)
+
+
+# from: https://github.com/huggingface/transformers/blob/v4.45.2/src/transformers/models/qwen2_vl/modeling_qwen2_vl.py
+class GaudiQwen2VisionSdpaAttention(VisionSdpaAttention):
+    def __init__(self, dim: int, num_heads: int = 16) -> None:
+        super().__init__(dim, num_heads)
+        self.fused_scaled_dot_product_attention = ModuleFusedSDPA(FusedSDPA) if FusedSDPA else None
+
+    def forward(
+        self,
+        hidden_states: torch.Tensor,
+        cu_seqlens: torch.Tensor,
+        rotary_pos_emb: torch.Tensor = None,
+        use_flash_attention: Optional[bool] = False,
+    ) -> torch.Tensor:
+        seq_length = hidden_states.shape[0]
+        q, k, v = self.qkv(hidden_states).reshape(seq_length, 3, self.num_heads, -1).permute(1, 0, 2, 3).unbind(0)
+        q = apply_rotary_pos_emb_vision(q.unsqueeze(0), rotary_pos_emb).squeeze(0)
+        k = apply_rotary_pos_emb_vision(k.unsqueeze(0), rotary_pos_emb).squeeze(0)
+
+        attention_mask = torch.zeros([1, seq_length, seq_length], device=q.device, dtype=torch.bool)
+        for i in range(1, len(cu_seqlens)):
+            attention_mask[:, cu_seqlens[i - 1] : cu_seqlens[i], cu_seqlens[i - 1] : cu_seqlens[i]] = True
+        q = q.transpose(0, 1)
+        k = k.transpose(0, 1)
+        v = v.transpose(0, 1)
+
+        if FusedSDPA is not None and use_flash_attention:
+            attn_output = self.fused_scaled_dot_product_attention(q, k, v, attention_mask, 0.0, False, None, "None")
+        else:
+            attn_output = F.scaled_dot_product_attention(q, k, v, attention_mask, dropout_p=0.0)
+
+        attn_output = attn_output.transpose(0, 1)
+        attn_output = attn_output.reshape(seq_length, -1)
+        attn_output = self.proj(attn_output)
+        del attention_mask
+        return attn_output
+
+
+GAUDI_QWEN2_VL_VISION_ATTENTION_CLASSES = {
+    "eager": VisionAttention,
+    "flash_attention_2": VisionFlashAttention2,
+    "sdpa": GaudiQwen2VisionSdpaAttention,
+}
```

Comment:
why do we need this? can you check how other model is done? only overwrite the function/class needed

Reply:
This was part of original code, now removed, only Sdpa+FusedSdpa attention is retained.